### PR TITLE
txn: fix `on_rollback` trigger arguments for transactions aborted by yield or timeout

### PR DESCRIPTION
**Pull request by Gumix**
#### txn: pass `txn_stmt` instead of `txn` to `on_commit`/`on_rollback`

Currently `on_rollback` triggers are called on rollback of the whole transaction. To make it possible to invoke them on rollback to a savepoint, we need to pass a statement at which the savepoint was created.

#### txn: fix `on_rollback` trigger args for txns aborted by yield/timeout

Currently, if a transaction is aborted by a fiber yield or by a timeout, `txn_rollback_to_svp()` is called to rollback all statements of the txn. After that the transaction is completely aborted on attempt to commit it. If an `on_rollback` trigger is set, it is called from `txn_complete_fail()`, however it receives an empty iterator, because at this point the statements are already destroyed. This patch invokes `on_rollback` triggers directly from `txn_rollback_to_svp()` for abort-by-yield, and postpones the rollback for abort-by-timeout.

#### txn: run `on_rollback` triggers on rollback to savepoint

Currently, if a statement is rolled back during rollback to a savepoint, it does not appear in neither `on_commit` nor `on_rollback` triggers. Fix this by running `on_rollback` triggers during the rollback to the savepoint.

Closes https://github.com/tarantool/tarantool/issues/7810
Closes https://github.com/tarantool/tarantool/issues/9340

### CODE REVIEW

File: changelogs/unreleased/gh-9340-fix-on_rollback-function-args.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+## bugfix/core
+
+* Fixed a bug when `on_rollback` trigger functions were invoked with an empty
+  iterator argument, if a transaction was aborted by a fiber yield or by a
+  timeout (gh-9340).
```
Comment (andreyaksenov) [inline changelogs/unreleased/gh-9340-fix-on_rollback-function-args.md:5]:
```suggestion
* Fixed a bug when `on_rollback` trigger functions were invoked with an empty
  iterator argument if a transaction was aborted by a fiber yield or by a
  timeout (gh-9340).
```

**Review by andreyaksenov — APPROVED** [review]

**Review by ochaplashkin — APPROVED** [review]

File: src/box/lua/init.c [hunk 2]
```diff
@@ -545,14 +545,14 @@ lbox_txn_iterator_next(struct lua_State *L)
...
-		return luaT_error(L);
-	}
 	struct txn_stmt *stmt =
 		(struct txn_stmt *) lua_topointer(L, lua_upvalueindex(2));
 	if (stmt == NULL)
 		return 0;
+	if (txn == NULL || txn->id != txn_id) {
+		diag_set(ClientError, ER_CURSOR_NO_TRANSACTION);
+		return luaT_error(L);
+	}
```
Comment (locker) [inline src/box/lua/init.c:555]: Why did you move this check? Looks like an unrelated change.

File: src/box/lua/init.c [hunk 3]
```diff
@@ -590,29 +590,33 @@ lbox_txn_pairs(struct lua_State *L)
...
  */
 static int
 lbox_push_txn(struct lua_State *L, void *event)
 {
-	struct txn *txn = (struct txn *) event;
-	luaL_pushint64(L, txn->id);
-	lua_pushcclosure(L, lbox_txn_pairs, 1);
+	struct txn_stmt *stmt = (struct txn_stmt *)event;
+	/* stmt is NULL for empty transactions. */
+	int64_t txn_id = stmt != NULL ? stmt->txn->id : -1;
```
Comment (locker) [inline src/box/lua/init.c:616]: Since `stmt` can be `NULL`, wouldn't it be better to use `in_txn()` here and in other triggers instead of `stmt->txn`?

File: changelogs/unreleased/gh-9340-fix-on_rollback-function-args.md (hunk 1 above)
Comment (Gumix) [reply changelogs/unreleased/gh-9340-fix-on_rollback-function-args.md]: Fixed.

File: src/box/lua/init.c (hunk 2 above)
Comment (Gumix) [reply src/box/lua/init.c]: Left over from previous version of the patch, fixed.

File: src/box/lua/init.c (hunk 3 above)
Comment (Gumix) [reply src/box/lua/init.c]: Done.

**Comment by Gumix** [comment]
Also changed an `if` to `txn_on_timeout` to optimize memory usage.

File: src/box/txn.c [hunk 4]
```diff
@@ -1497,10 +1523,16 @@ txn_on_yield(struct trigger *trigger, void *event)
...
+		txn_rollback_to_svp(txn, NULL, true);
 		txn->status = TXN_ABORTED;
 		txn_set_flags(txn, TXN_IS_ABORTED_BY_YIELD);
 		say_warn("Transaction has been aborted by a fiber yield");
+		if (txn_has_flag(txn, TXN_HAS_TRIGGERS)) {
+			/* Can't rollback more than once. */
+			trigger_destroy(&txn->on_rollback);
+			/* Commit won't happen after rollback. */
+			trigger_destroy(&txn->on_commit);
+		}
```
Comment (Gerold103) [inline src/box/txn.c:1535]: From that moment you have the destroyed triggers here but the flag `TXN_HAS_TRIGGERS` is still set. Either the triggers need to be re-created empty, or the flag needs to be unset.

**Review by Gerold103 — APPROVED** [review]
Thanks for the patch!

File: src/box/txn.c (hunk 4 above)
Comment (locker) [reply src/box/txn.c:1535]: AFAIU after `trigger_destroy` is called, the trigger list becomes empty, but yeah, it's probably better to reinitialize it explicitly.
